proposals/0047-syscall-to-get-last-restart-slot.md

```diff
@@ -0,0 +1,113 @@
+---
+simd: '0047'
+title: Syscall to get the last restart slot
+authors:
+  - Godmode Galactus (Mango Markets)
+category: Standard
+type: Core
+status: Draft
+created: 2023-04-15
+---
+
+## Summary
+
+Create a new syscall that can be used to get the last restart slot (currently
+the last time hard fork was done on the cluster).
+
+`fn sol_get_last_restart_slot() -> Slot`
+
+## Motivation
+
+In Solana, when the cluster cannot reach consensus, it is currently restarted
+using `ledger-tool` with a hard fork on a slot. All participating nodes need to
+restart their validator specifying the hard fork slot. Once all nodes
+participating in the restart effort on the hard fork exceed 80% of the total
+stakes, the restart is successful, and the cluster continues from the hard fork
+slot. So hard fork is a tool to intentionally fork off the nodes not
+participating in the restart effort. Currently, we can consider the slot on
+which hard fork was done as a restart slot.
+
+Program developers may find it useful to know that a hard fork has recently
+occurred. This information can help them prevent arbitrage and liquidation
+caused by outdated oracle price account states. However, the cluster's restart
+process takes several hours; in that time, the world can change, causing asset
+prices to fluctuate. As the cluster updates the state of all accounts, malicious
+actors could take advantage of the delay by conducting trades using outdated
+prices, anticipating that they will soon catch up to the actual prices of
+real-world assets. Knowing that cluster restart has been done recently, programs
+can manage these cases more appropriately.
+
+## Alternatives Considered
+
+We need to have the value of the last restart slot while executing the program.
+We cannot use an account because then it should be updated just after the
+restart is successful, which will add complexity. The best way is to create a
+new syscall or sysvar to get this information during the execution of a
+transaction, which will help us get the last restart slot. We prefer syscall
+over sysvar because the newly created sysvar has to be passed in the instruction
+this takes up transaction accounts space. This will break existing program
+interfaces, and the transaction account space is limited which is already a pain
+point for many DeFi projects.
```

Comment:
block time vs slot increase: when cluster time jumps by more than 1m between two slots, there was most likely a restart a few slots ago. issue is that this change is slightly delayed as it needs a regular stake majority to vote after the restart.

Reply:
Here is an example from a recent restart:

403 was the fork
404 was still a vob
405 had user transactions
406 the cluster clock was updated